protected/tests/unit/models/AUSPatientTest.php

```diff
@@ -18,45 +18,244 @@ public function tearDown()
         parent::tearDown();
     }
 
-    public function testUniqueMedicareNo()
+
+    //create a new patient model with everything in order
+    protected function getUnsavedNewPatient($patient_key)
+    {
+        $new_pat = new Patient;
+        $new_pat->attributes = $this->patient[$patient_key];
+        $new_pat->id = NULL;
+        $new_pat->pas_key = NULL;
+        $new_pat->nhs_num = $this->getNDigitRand();
+        $new_pat->hos_num = $this->getNDigitRand();
+
+        return $new_pat;
+    }
+
+    //create a GP with everything in order
+    protected function getGP()
+    {
+        $new_gp = new Gp();
+        $new_gp->obj_prof = $this->getNDigitRand(20);
+        $new_gp->nat_id = $this->getNDigitRand(20);
+        $new_gp->save();
+        return $new_gp;
+    }
+
+    protected function getNDigitRand($noDigits = 40)
     {
-        //create a patient as a duplicate of one in the database
-        $patient_dup = new Patient;
-        $patient_dup->attributes = $this->patient['patient1'];
+        $rand_string = mt_rand(1,9);
+        for ($i = 1; $i < $noDigits; $i++)
+        {
+            $rand_string .= mt_rand(0,9);
+        }
+        return $rand_string;
+    }
 
-        //change the things we don't want it to fail on
-        $patient_dup->id = NULL;
-        $patient_dup->pas_key = NULL;
-        $patient_dup->hos_num = 65465465;
+    public function testUniqueMedicareNo()
+    {
+        $patient_dup = $this->getUnsavedNewPatient('patient1');
+        $patient_dup->nhs_num = $this->patient['patient1']['nhs_num'];
 
         //fail to save with duplication
         $this->assertFalse($patient_dup->save());
         $this->assertSame('Medicare Number "" has already been taken.' //remove the actual number
             , preg_replace('/\d+/', '' , $patient_dup->getErrors('nhs_num')[0]));
 
         //remove duplication and save successfully
-        $patient_dup->nhs_num = 65465445;
+        $patient_dup->nhs_num = $this->getNDigitRand();
```

Comment:
Won't this collide once every 1.e+40 times? Is that an acceptable failure rate?

Reply:
The database gets cleaned down after each unit test runs, so collisions would be extremely rare. Re-running the unit test would allow it to pass.